pkg/chunk/local/archiver.go

```diff
@@ -0,0 +1,435 @@
+	AWSStorageConfig aws.StorageConfig `yaml:"aws"`
+	GCSConfig        gcp.GCSConfig     `yaml:"gcs"`
+	FSConfig         FSConfig          `yaml:"filesystem"`
+}
+
+// ArchiverConfig is for configuring an archiver
+type ArchiverConfig struct {
+	CacheLocation  string        `yaml:"cache_location"`
+	CacheTTL       time.Duration `yaml:"cache_ttl"`
+	StoreConfig    StoreConfig   `yaml:"store_config"`
+	ResyncInterval time.Duration `yaml:"resync_interval"`
+	IngesterName   string        `yaml:"-"`
+	Mode           int           `yaml:"-"`
+}
+
+// ArchiveStoreClient define all the methods that a store needs to implement for managing objects
+type ArchiveStoreClient interface {
+	GetObject(ctx context.Context, objectName string) ([]byte, error)
+	PutObject(ctx context.Context, objectName string, object []byte) error
+	List(ctx context.Context, prefix string) (map[string]time.Time, error)
+}
+
+// RegisterFlags registers flags.
+func (cfg *ArchiverConfig) RegisterFlags(f *flag.FlagSet) {
+	storeFlagsPrefix := "boltdb.archiver."
+	cfg.StoreConfig.AWSStorageConfig.RegisterFlagsWithPrefix(storeFlagsPrefix, f)
+	cfg.StoreConfig.GCSConfig.RegisterFlagsWithPrefix(storeFlagsPrefix, f)
+	cfg.StoreConfig.FSConfig.RegisterFlagsWithPrefix(storeFlagsPrefix, f)
+
+	f.StringVar(&cfg.StoreConfig.Store, "boltdb.archiver.store", "filesystem", "Store for keeping boltdb files")
+	f.StringVar(&cfg.CacheLocation, "boltdb.archiver.cache-location", "", "Cache location for restoring boltDB files for queries")
+	f.DurationVar(&cfg.CacheTTL, "boltdb.archiver.cache-ttl", 24*time.Hour, "TTL for boltDB files restored in cache for queries")
+	f.DurationVar(&cfg.ResyncInterval, "boltdb.archiver.resync-interval", 5*time.Minute, "Resync downloaded files with the store")
+}
+
+// Update holds a single update to a file or folder
+type Update struct {
+	UpdateType int
+	TableName  string
+	FilePath   string
+}
+
+// downloadedTable represents multiple boltdb files for same period uploaded by different ingesters
+type downloadedTable struct {
+	files        map[string]time.Time
+	downloadedAt time.Time
+}
+
+// Archiver holds its configuration and progress of objects in sync
+// Uploads boltdb files to storage with structure <boltdb-filename>/<ingester-name>
+// Keeps syncing locally changed file to the storage and downloading latest changes from storage
+// Cleans up downloaded files as per configured TTL
+type Archiver struct {
+	cfg             ArchiverConfig
+	boltDbDirectory string
+	storeClient     ArchiveStoreClient
+
+	downloadedTables    map[string]*downloadedTable
+	downloadedTablesMtx sync.RWMutex
+
+	updatesChan           chan Update
+	uploadedFilesMtime    map[string]time.Time
+	uploadedFilesMtimeMtx sync.RWMutex
+	done                  chan struct{}
+
+	// We would use ingester name and startup timestamp for naming files while uploading so that
+	// ingester does not override old files when using same id
+	ingesterNameAndStartUpTs string
```

Comment:
We may rename it to something like `ingesterUniqueID` (the fact it's built upon ingester name + timestamp is an implementation detail, but what we need to store in the variable is a unique ID for the ingester).